### DISCUSSION TITLE
What's with "Enumerable.Reverse()"?

### DISCUSSION THREAD

**Discussion by VladiStep** [dotnet/runtime#111121] [General]
The `IEnumerable.Reverse()` always does buffer the items, even for lists and arrays.
And also the buffering is not mentioned in the docs (e.g. https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.reverse).

There is an issue #15826, but it was 9 years ago, and I want to know if the opinion is the same.

**Comment by VladiStep** [comment]
Also, why "SourceLink" ("Go to implementation" option in Visual Studio) says that it uses `Buffer<TSource>` when [the source code here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Reverse.cs) says it uses just an array (`TSource[]`) ?

![Image](https://github.com/user-attachments/assets/4c3dbca9-48c8-46f3-b120-9beffcd21efa)

**Comment by VladiStep** [comment]
I almost forgot - it's .NET 6 ("C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.33\System.Linq.dll").
But it seems that the "release/6.0" branch has the same code for `ReverseIterator` as the "main".

**Comment by VladiStep** [comment]
https://github.com/dotnet/runtime/issues/15826#issuecomment-161162038
![Image](https://github.com/user-attachments/assets/d3ddeb85-9220-4b31-a7bc-565ce95e76d0)

> \- you should document this
> \- nah it's fine people can find it by decompiling

😕

**Comment by huoyaoyuan** [comment]
Look carefully, the 6.0 version does use `Buffer<TSource>`: https://github.com/dotnet/runtime/blob/release/6.0/src/libraries/System.Linq/src/System/Linq/Reverse.cs#L57

**Comment by VladiStep** [comment]
> Look carefully, the 6.0 version does use `Buffer<TSource>`: https://github.com/dotnet/runtime/blob/release/6.0/src/libraries/System.Linq/src/System/Linq/Reverse.cs#L57

Hm, I opened the branch and searched for `ReverseIterator` (using the search bar), but it seems it didn't open the right one.
Is the search bar broken?

**Comment by vcsjones** [comment]
> Is the search bar broken?

If you mean the GitHub search, GitHub search is currently limited to the `main` branch.

**Comment by VladiStep** [comment]
@eiriktsarpalis, so the opinion is the same?

**Comment by eiriktsarpalis** [comment]
Not sure why I closed this, reopening.
